src/keystore/keystore.ts

```diff
@@ -0,0 +1,237 @@
+    membershipHash: MembershipHash,
+    password: Password
+  ): Promise<null | Uint8Array> {
+    const nwakuCredential = this.data.credentials[membershipHash];
+
+    if (!nwakuCredential) {
+      return null;
+    }
+
+    const eipKeystore = Keystore.fromCredentialToEip(nwakuCredential);
+    return decryptEipKeystore(eipKeystore, password);
+  }
+
+  public removeCredential(hash: MembershipHash): void {
+    if (!this.data.credentials[hash]) {
+      return;
+    }
+
+    delete this.data.credentials[hash];
+  }
+
+  public toString(): string {
+    return JSON.stringify(this.data);
+  }
+
+  private static isValidNwakuStore(obj: unknown): boolean {
+    if (!isKeystoreValid(obj)) {
+      return false;
+    }
+
+    const areCredentialsValid = Object.values(_.get(obj, "credentials", {}))
+      .map((c) => isCredentialValid(c))
+      .every((v) => v);
+
+    return areCredentialsValid;
+  }
+
+  private static fromCredentialToEip(
+    credential: NwakuCredential
+  ): IEipKeystore {
+    const nwakuCrypto = credential.crypto;
+    const eipCrypto: IEipKeystore["crypto"] = {
+      kdf: {
+        function: nwakuCrypto.kdf,
+        params: nwakuCrypto.kdfparams,
+        message: "",
+      },
+      cipher: {
+        function: nwakuCrypto.cipher,
+        params: nwakuCrypto.cipherparams,
+        message: nwakuCrypto.ciphertext,
+      },
+      checksum: {
+        function: "sha256",
+        params: {},
+        message: nwakuCrypto.mac,
+      },
+    };
+
+    return {
+      version: 4,
+      uuid: uuidV4(),
+      description: undefined,
+      path: "safe to ignore, not important for decrypt",
+      pubkey: "safe to ignore, not important for decrypt",
+      crypto: eipCrypto,
+    };
+  }
+
+  private static fromEipToCredential(
+    eipKeystore: IEipKeystore
+  ): NwakuCredential {
+    const eipCrypto = eipKeystore.crypto;
+    const eipKdf = eipCrypto.kdf as IPbkdf2KdfModule;
+    return {
+      crypto: {
+        cipher: eipCrypto.cipher.function,
+        cipherparams: eipCrypto.cipher.params,
+        ciphertext: eipCrypto.cipher.message,
+        kdf: eipKdf.function,
+        kdfparams: eipKdf.params,
+        mac: eipCrypto.checksum.message,
+      },
+    };
+  }
+
+  // follows nwaku implementation
+  // https://github.com/waku-org/nwaku/blob/f05528d4be3d3c876a8b07f9bb7dfaae8aa8ec6e/waku/waku_keystore/protocol_types.nim#L111
+  private static computeMembershipHash(info: MembershipInfo): string {
+    return bytesToUtf8(
+      sha256(utf8ToBytes(`${info.chainId}${info.address}${info.treeIndex}`))
+    );
+  }
+
+  // follows nwaku implementation
+  // https://github.com/waku-org/nwaku/blob/f05528d4be3d3c876a8b07f9bb7dfaae8aa8ec6e/waku/waku_keystore/protocol_types.nim#L98
+  private static computeIdentityToBytes(
+    identity: IdentityCredential,
+    membershipInfo: MembershipInfo
+  ): Uint8Array {
+    // TODO: fix implementation once clarified
+    const str = `${JSON.stringify(identity)}=${JSON.stringify(
```

Comment:
@rymnc & @alrevuelta please, correct me here if I misunderstand something.

Does `nwaku` encrypts [following sting](https://github.com/waku-org/nwaku/blob/2c5eb427d7a15cafbee50fd7b97a39be6e6347ce/waku/waku_keystore/protocol_types.nim#L98) into `ciphertext`? 

If so how it then parses it? Or it encrypts stringified representation of `KeystoreMembership`?

Reply:
yes, maybe it will make more sense with this example - https://github.com/waku-org/nwaku/blob/master/waku/waku_keystore/keyfile.nim#L332-L395

Reply:
it's a stringified version of KeystoreMembership

Reply:
@rymnc, then I am wondering how does it get's re-exported into `nwaku`? Is there some parser that takes string `"KeystoreMembership(...)"` and builds `KeystoreMembership` object from it?

Reply:
https://github.com/waku-org/nwaku/blob/master/waku/waku_keystore/conversion_utils.nim#L12-L31 here we describe it

Reply:
it should be compatible now